Implement TextMesh.

bruvzg opened the pull request:
Follow up to the #60386

Implement `TextMesh` resource.

Implements mesh generation from the shaped text contours. Works only with the OpenType compliant dynamic fonts (without contour self-intersections, so some fonts converted by services like Google Font might fail). Will not work with bitmap fonts (regardless of font container format), e.g., color emoji fonts.

TODO:

- [x] Optimize generated mesh.
- [x] Generate a reasonable texture coordinates for the mesh.
- [x] Implement fallbacks for missing and faulty glyphs.

<img width="730" alt="Screenshot 2022-04-26 at 13 43 52" src="https://user-images.githubusercontent.com/7645683/165283198-6990686a-0dea-42c1-95af-6004bcc943e6.png">
<img width="730" alt="Screenshot 2022-04-26 at 13 43 57" src="https://user-images.githubusercontent.com/7645683/165283207-c99ca433-9878-430a-be73-5142efa80863.png">
<img width="730" alt="Screenshot 2022-04-26 at 23 25 06" src="https://user-images.githubusercontent.com/7645683/165386132-d3403d21-1439-4a5f-9c96-e2454ef83f53.png">

Test project:
[test_project.zip](https://github.com/godotengine/godot/files/8569863/test_project.zip)

bruvzg committed:
Implement TextMesh resource.

Apply simulated slant and embolden to the TextServer `gont_get_glyph_contours` results.

fire commented:
Can you send it through the lod system?

Calinou commented:
I gave it a try :slightly_smiling_face: 

![image](https://user-images.githubusercontent.com/180032/165174020-44752b28-60c8-4bca-b973-c9b7c03b3d09.png)

**Testing project:** [test_textmesh.zip](https://github.com/godotengine/godot/files/8557933/test_textmesh.zip)

Some comments:

- Text is left-aligned by default, as alignment options aren't implemented yet.
- Spaces aren't rendered yet.
- The mesh fails to generate with an error message if you edit the `text` property to make it an empty string again. Due to this, the last character that was present in the `text` property will also remain visible.
- Generating the mesh can take a lot of time, which can cause stuttering if the text changes during gameplay. The longer the text, the more noticeable this is. Increasing **Curve Step** in the TextMesh resource can improve mesh generation performance a lot (at the cost of worse curve quality).
  - I'm not sure if we can fully resolve this issue (given that uploading new meshes to the GPU will always take some time). Caching individual glyph triangles for each font might help a little.
  - For people looking to create 3D timers in the game world, we can probably document a way to do this with several MeshInstance3D nodes. This can be done by creating 10 MeshInstance3D nodes per digit (with all numbers in advance), and toggling the nodes' visibility at run-time. This should avoid performance caveats related to run-time mesh generation.
- As an optimization, generating triangles on the side of the text can be skipped if `depth` is nearly equal to 0.

We should also look to compile some use cases for TextMesh where Label3D isn't a good fit. Label3D is a powerful node already, and we should be careful not to have feature creep in this aspect of the engine.

anderlli0053 commented:
Can this be ported to 3.x?

bruvzg commented:
- Added glyph data caching (local only), it should be much faster (the slowest part is convex decomposition). Now mesh generation in the `TextMesh` should account only 10% of the mesh update time (Another 10% is taken by base `Mesh` update, and the rest by the `RenderingServer`).
- Added alignment property.
- Empty string and spaces should be fixed, and missing glyph replaced with quads.
- Added property tags for curve step and depth.
- Removed side and back faces generation when depth is zero.

> Aka. The Meshoptimizer system? I can look into this later.

Not sure if it's necessary, with the reasonable curve step it should be good as is.

> If you can't generate the uvs manually, you could use xatlas.

It's probably a bit of overkill to use full unwrap, something simple and applicable to any text without changing the texture probably is a more useful as a default and should cover most cases. So currently it's using a map like this to have a possibility to texture front back and sides of text independently (front and back and for the full text, and sides to loop each contour):

<img width="730" alt="Screenshot 2022-04-26 at 23 25 06" src="https://user-images.githubusercontent.com/7645683/165386132-d3403d21-1439-4a5f-9c96-e2454ef83f53.png">

But I guess adding a separate option to do unwrap on any mesh would be useful.

bruvzg commented:
> Can this be ported to 3.x?

Probably not, at least not without backporting some other stuff to the 3.x, and it won't have any of the TextServer features available.

Calinou commented:
I tested the latest revision of this PR and can confirm the timer in the testing project now performs much better (even when it uses 4 decimals).

golddotasksquestions commented:
I just spend at least week on painstakingly trying to get dynamic 3D text mesh (mesh created in Blender letter by letter) in one of my projects in Godot 3 to write 4 words. It's super limited and I can't imagine doing this ever again and was about to write a feature request proposal of exactly this!

You are amazing!

Question: Will I be able to generate CollisionShape Siblings from this like I can with MeshInstances (or alternatively make this TextMesh the Mesh of a MeshInstance so I can generate the CollisionShape Siblings)?

bruvzg commented:
> Question: Will I be able to generate CollisionShape Siblings from this like I can with MeshInstances (or alternatively make this TextMesh the Mesh of a MeshInstance so I can generate the CollisionShape Siblings)?

I have not tested collisions, but it's a derivative of generic `PrimitiveMesh` not a full separate `Node`, so it can be used `mesh` of the `MeshInstance3D` (that's what happening in the test project), and should support everything any other mesh do.

akien-mga reviewed (approved):
The feature looks great, and the code seems fine from a quick look. Might warrant a rebase to make sure it still builds after the HashMap rewrite.

timothyqiu commented:
The "Pixel Size" property somehow updates its character spacing in the editor preview.

It works fine if I reopen the scene.

![ksnip_20220513-103547](https://user-images.githubusercontent.com/372476/168200251-84138b32-7bf2-4382-94d1-696be8cbab09.png)

timothyqiu commented on scene/resources/primitive_meshes.cpp:
```diff
@@ -2151,3 +2156,798 @@ void RibbonTrailMesh::_bind_methods() {
+	ADD_PROPERTY(PropertyInfo(Variant::STRING, "text"), "set_text", "get_text");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "font", PROPERTY_HINT_RESOURCE_TYPE, "Font"), "set_font", "get_font");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "font_size", PROPERTY_HINT_RANGE, "1,127,1"), "set_font_size", "get_font_size");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "horizontal_alignment", PROPERTY_HINT_ENUM, "Left,Center,Right,Fill"), "set_horizontal_alignment", "get_horizontal_alignment");
```
Since the documentation says only left, center, and right are supported, maybe we should remove the "Fill" option here?

bruvzg replied:
Fixed `pixel_size` not cleaning cache, added `width` for `Fill` alignment and updated docs for this and `Label3D` (both support `Fill` alignment).

akien-mga commented:
Thanks!
